lib/corefoundation/base_wrapper.rb

```diff
@@ -0,0 +1,112 @@
+# The top level namespace for the corefoundation library. The raw FFI generated methods are attached here
+module CF
+  extend FFI::Library
+  ffi_lib "/System/Library/Frameworks/CoreFoundation.framework/CoreFoundation"
+
+  if FFI::Platform::ARCH == "x86_64"
+    typedef :long_long, :cfindex
+    typedef :long_long, :cfcomparisonresult
+    typedef :ulong_long, :cfoptionflags
+    typedef :ulong_long, :cftypeid
+    typedef :ulong_long, :cfhashcode
+  else
+    typedef :long, :cfindex
+    typedef :long, :cfcomparisonresult
+    typedef :ulong, :cfoptionflags
+    typedef :ulong, :cftypeid
+    typedef :ulong, :cfhashcode
+  end
+
+  typedef :pointer, :cftyperef
+
+  attach_function :show, "CFShow", [:cftyperef], :void
+  attach_function :release, "CFRelease", [:cftyperef], :void
+  attach_function :retain, "CFRetain", [:cftyperef], :cftyperef
+  attach_function "CFEqual", %i{cftyperef cftyperef}, :char
+  attach_function "CFHash", [:cftyperef], :cfhashcode
+  attach_function "CFCopyDescription", [:cftyperef], :cftyperef
+  attach_function "CFGetTypeID", [:cftyperef], :cftypeid
+
+  # @private
+  class Range < FFI::Struct
+    layout :location, :cfindex,
+           :length, :cfindex
+  end
+
+  # Provides a more declarative way to define all required abstract methods.
+  # This gets included in the CF::Base class whose subclasses then need to define those methods.
+  module BaseWrapper
+    @registered_types = {}
+
+    def self.included(base)
+      base.extend RegisteredTypes
+      base.class_eval do
+        def self.abstract_methods(*methods)
+          methods.each do |name|
+            define_method name do
```

Comment:
As a whole, this module seems to be using some metaprogramming and non-idiomatic ruby. This will create problem in terms of maintainability and debugging. Since, we already have considerable complexity interfacing with CF types, can we try simplifying this abstraction/module?

Reply:
How do you suggest we go about this? My goal in doing this is to force the methods to be defined in subclasses and throw error when they don't have them. Since the previous approach didn't actually enforce the methods to be defined in subclasses which made it possible to overlook. It was more of a "development meta knowledge" - which didn't feel like the right thing to do.

Reply:
> My goal in doing this is to force the methods to be defined in subclasses and throw error when they don't have them

Thought, we usually don't have abstract methods in ruby, https://stackoverflow.com/a/6792499 but still if it seems a good idea here, let's have a separate module for that instead of putting in `BaseWrapper`

Reply:
So something like a `BaseAbstract` to specifically enforce a set of functions/provide defaults and this module would only contain memory management related helpers then?

Reply:
Will be doing some changes around this on a separate PR. related comment: https://github.com/chef/corefoundation/pull/10#issuecomment-894164089 Resolving this conversation for now.

Reply:
Could you revert these code changes then?